File: lincheck/src/main/java/org/jetbrains/kotlinx/lincheck/LinChecker.java
package org.jetbrains.kotlinx.lincheck;

/*
 * #%L
 * Lincheck
 * %%
 * Copyright (C) 2015 - 2018 Devexperts, LLC
 * %%
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Lesser Public License for more details.
 *
 * You should have received a copy of the GNU General Lesser Public
 * License along with this program.  If not, see
 * <http://www.gnu.org/licenses/lgpl-3.0.html>.
 * #L%
 */

import org.jetbrains.kotlinx.lincheck.annotations.LogLevel;
import org.jetbrains.kotlinx.lincheck.execution.*;
import org.jetbrains.kotlinx.lincheck.strategy.Strategy;
import org.jetbrains.kotlinx.lincheck.verifier.*;
import static org.jetbrains.kotlinx.lincheck.ReporterKt.DEFAULT_LOG_LEVEL;
import java.util.*;


/**
 * This class runs concurrent tests.
 * See {@link #check(Class)} and {@link #check(Class, Options)} methods for details.
 */
public class LinChecker {
    private final Class<?> testClass;
    private final List<? extends CTestConfiguration> testConfigurations;
    private final CTestStructure testStructure;
    private final Reporter reporter;

    private LinChecker(Class<?> testClass, Options options) {
        this.testClass = testClass;
        this.testStructure = CTestStructure.getFromTestClass(testClass);
        LoggingLevel logLevel;
        if (options != null) {
            logLevel = options.logLevel;
            this.testConfigurations = Collections.singletonList(options.createTestConfigurations(testClass));
        } else {
            logLevel = getLogLevelFromAnnotation();
            this.testConfigurations = CTestConfiguration.createFromTestClass(testClass);
        }
        this.reporter = new Reporter(logLevel);
    }

    /**
     * Runs all concurrent tests described with {@code @<XXX>CTest} annotations on the specified test class.
     *
     * @throws AssertionError if algorithm or data structure is not correct.
     */
    public static void check(Class<?> testClass) {
        check(testClass, null);
    }

    /**
     * Runs concurrent test on specified class with the specified by options environment.
     * <p>
     * NOTE: this method ignores {@code @<XXX>CTest} annotations on the test class.
     *
     * @throws AssertionError if algorithm or data structure is not correct.
     */
    public static void check(Class<?> testClass, Options options) {
        new LinChecker(testClass, options).check();
    }

    /**
     * @throws AssertionError if algorithm or data structure is not correct
     */
    private void check() throws AssertionError {
        if (testConfigurations.isEmpty()) {
            throw new IllegalStateException("No Lin-Check test configuration to run");
        }
        testConfigurations.forEach(testConfiguration -> {
            try {
                checkImpl(testConfiguration);
            } catch (RuntimeException | AssertionError e) {
                throw e;
            } catch (Exception e) {
                throw new IllegalStateException(e);
            }
        });
    }

    private void checkImpl(CTestConfiguration testCfg) throws AssertionError, Exception {
        ExecutionGenerator exGen = createExecutionGenerator(testCfg.generatorClass, testCfg);
        // Run iterations
        for (int iteration = 1; iteration <= testCfg.iterations; iteration++) {
            ExecutionScenario scenario = exGen.nextExecution();
            // Check correctness of the generated scenario
            reporter.logIteration(iteration, testCfg.iterations, scenario);
            try {
                runScenario(scenario, testCfg);
            } catch (AssertionError e) {
                if (!testCfg.minimizeFailedScenario) throw e;
                minimizeScenario(scenario, testCfg, e);
            }
        }
    }

    // Tries to minimize the specified failing scenario to make the error easier to understand.
    // The algorithm is greedy: it tries to remove one actor from the scenario and checks
    // whether a test with the modified one fails with error as well. If it fails,
    // then the scenario has been successfully minimized, and the algorithm tries to minimize it again, recursively.
    // Otherwise, if no actor can be removed so that the generated test fails, the minimization is completed.
    // Thus, the algorithm works in the linear time of the total number of actors.
    private void minimizeScenario(ExecutionScenario scenario, CTestConfiguration testCfg, AssertionError currentError) throws AssertionError, Exception {
        reporter.logScenarioMinimization(scenario);
        for (int i = 0; i < scenario.parallelExecution.size(); i++) {
            for (int j = 0; j < scenario.parallelExecution.get(i).size(); j++) {
                ExecutionScenario newScenario = copyScenario(scenario);
                newScenario.parallelExecution.get(i).remove(j);
                if (newScenario.parallelExecution.get(i).isEmpty())
                    newScenario.parallelExecution.remove(i); // remove empty thread
                minimizeNewScenarioAttempt(newScenario, testCfg);
            }
        }
        for (int i = 0; i < scenario.initExecution.size(); i++) {
            ExecutionScenario newScenario = copyScenario(scenario);
            newScenario.initExecution.remove(i);
            minimizeNewScenarioAttempt(newScenario, testCfg);
        }
        for (int i = 0; i < scenario.postExecution.size(); i++) {
            ExecutionScenario newScenario = copyScenario(scenario);
            newScenario.postExecution.remove(i);
            minimizeNewScenarioAttempt(newScenario, testCfg);
        }
        throw currentError;
    }

    private void minimizeNewScenarioAttempt(ExecutionScenario newScenario, CTestConfiguration testCfg) throws AssertionError, Exception {
        try {
            runScenario(newScenario, testCfg);
        } catch (IllegalArgumentException e) {
            // Ignore incorrect scenarios
        } catch (AssertionError e) {
            // Scenario has been minimized! Try to minimize more!
            minimizeScenario(newScenario, testCfg, e);
            throw new IllegalStateException("Never reaches, the previous `minimizeScenario` call always throw an exception");
        }
    }

    private ExecutionScenario copyScenario(ExecutionScenario scenario) {
        List<Actor> initExecution = new ArrayList<>(scenario.initExecution);
        List<List<Actor>> parallelExecution = new ArrayList<>();
        for (int i = 0; i < scenario.parallelExecution.size(); i++) {
            parallelExecution.add(new ArrayList<>(scenario.parallelExecution.get(i)));
        }
        List<Actor> postExecution = new ArrayList<>(scenario.postExecution);
        return new ExecutionScenario(initExecution, parallelExecution, postExecution);
    }

    private void runScenario(ExecutionScenario scenario, CTestConfiguration testCfg) throws AssertionError, Exception {
        validateScenario(testCfg, scenario);
        Verifier verifier = createVerifier(testCfg.verifierClass, testCfg.sequentialSpecification);
        if (testCfg.requireStateEquivalenceImplCheck) verifier.checkStateEquivalenceImplementation();
        Strategy strategy = Strategy.createStrategy(testCfg, testClass, scenario, verifier, reporter);
        strategy.run();
    }

    private void validateScenario(CTestConfiguration testCfg, ExecutionScenario scenario) {
        if (scenario.hasSuspendableActors()) {
            if (scenario.initExecution.stream().anyMatch(Actor::isSuspendable))
                throw new IllegalArgumentException("Generated execution scenario for the test class with suspendable methods contains suspendable actors in initial part");
            if (scenario.parallelExecution.stream().anyMatch(actors -> actors.stream().anyMatch(Actor::isSuspendable)) && scenario.postExecution.size() > 0)
                throw new IllegalArgumentException("Generated execution scenario for the test class with suspendable methods has non-empty post part");
        }
    }

    private Verifier createVerifier(Class<? extends Verifier> verifierClass, Class<?> sequentialSpecification) throws Exception {
        return verifierClass.getConstructor(Class.class).newInstance(sequentialSpecification);
    }

    private ExecutionGenerator createExecutionGenerator(Class<? extends ExecutionGenerator> generatorClass,
                                                        CTestConfiguration testConfiguration) throws Exception {
        return generatorClass.getConstructor(CTestConfiguration.class, CTestStructure.class).newInstance(testConfiguration, testStructure);
    }

    private LoggingLevel getLogLevelFromAnnotation() {
        LogLevel logLevelAnn = testClass.getAnnotation(LogLevel.class);
        if (logLevelAnn == null)
            return DEFAULT_LOG_LEVEL;
        return logLevelAnn.value();
    }
}

File: src/main/java/org/jetbrains/kotlinx/lincheck/LinChecker.kt
/*
 * #%L
 * Lincheck
 * %%
 * Copyright (C) 2019 JetBrains s.r.o.
 * %%
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Lesser Public License for more details.
 *
 * You should have received a copy of the GNU General Lesser Public
 * License along with this program.  If not, see
 * <http://www.gnu.org/licenses/lgpl-3.0.html>.
 * #L%
 */
package org.jetbrains.kotlinx.lincheck

import org.jetbrains.kotlinx.lincheck.CTestConfiguration.*
import org.jetbrains.kotlinx.lincheck.annotations.*
import org.jetbrains.kotlinx.lincheck.execution.*
import org.jetbrains.kotlinx.lincheck.strategy.*
import org.jetbrains.kotlinx.lincheck.verifier.*

/**
 * This class runs concurrent tests.
 */
class LinChecker (private val testClass: Class<*>, options: Options<*, *>?) {
    private val testStructure = CTestStructure.getFromTestClass(testClass)
    private val testConfigurations: List<CTestConfiguration>
    private val reporter: Reporter

    init {
        val logLevel = options?.logLevel ?: testClass.getAnnotation(LogLevel::class.java)?.value ?: DEFAULT_LOG_LEVEL
        reporter = Reporter(logLevel)
        testConfigurations = if (options != null) listOf(options.createTestConfigurations(testClass))
                             else createFromTestClassAnnotations(testClass)
    }

    /**
     * @throws LincheckAssertionError if the testing data structure is incorrect.
     */
    fun check() {
        val failure = checkImpl() ?: return
        throw LincheckAssertionError(failure)
    }

    /**
     * @return TestReport with information about concurrent test run.
     */
    internal fun checkImpl(): LincheckFailure? {
        check(testConfigurations.isNotEmpty()) { "No Lincheck test configuration to run" }
        for (testCfg in testConfigurations) {
            val failure = testCfg.checkImpl()
            if (failure != null) return failure
        }
        return null
    }

    private fun CTestConfiguration.checkImpl(): LincheckFailure? {
        val exGen = createExecutionGenerator()
        val verifier = createVerifier()
        repeat(iterations) { iteration ->
            val scenario = exGen.nextExecution()
            scenario.validate()
            reporter.logIteration(iteration, iterations, scenario)
            val failure = scenario.run(this, verifier)
            if (failure != null) {
                val minimizedFailedIteration = if (!minimizeFailedScenario) failure
                                               else failure.minimize(this, verifier)
                reporter.logFailedIteration(minimizedFailedIteration)
                return minimizedFailedIteration
            }
        }
        return null
    }

    // Tries to minimize the specified failing scenario to make the error easier to understand.
    // The algorithm is greedy: it tries to remove one actor from the scenario and checks
    // whether a test with the modified one fails with error as well. If it fails,
    // then the scenario has been successfully minimized, and the algorithm tries to minimize it again, recursively.
    // Otherwise, if no actor can be removed so that the generated test fails, the minimization is completed.
    // Thus, the algorithm works in the linear time of the total number of actors.
    private fun LincheckFailure.minimize(testCfg: CTestConfiguration, verifier: Verifier): LincheckFailure {
        reporter.logScenarioMinimization(scenario)
        for (i in scenario.parallelExecution.indices) {
            for (j in scenario.parallelExecution[i].indices) {
                val newScenario = scenario.copy()
                newScenario.parallelExecution[i].removeAt(j)
                if (newScenario.parallelExecution[i].isEmpty()) newScenario.parallelExecution.removeAt(i) // remove empty thread
                val newFailedIteration = newScenario.tryMinimize(testCfg, verifier)
                if (newFailedIteration != null) return newFailedIteration.minimize(testCfg, verifier)
            }
        }
        for (i in scenario.initExecution.indices) {
            val newScenario = scenario.copy()
            newScenario.initExecution.removeAt(i)
            val newFailedIteration = newScenario.tryMinimize(testCfg, verifier)
            if (newFailedIteration != null) return newFailedIteration.minimize(testCfg, verifier)
        }
        for (i in scenario.postExecution.indices) {
            val newScenario = scenario.copy()
            newScenario.postExecution.removeAt(i)
            val newFailedIteration = newScenario.tryMinimize(testCfg, verifier)
            if (newFailedIteration != null) return newFailedIteration.minimize(testCfg, verifier)
        }
        return this
    }

    private fun ExecutionScenario.tryMinimize(testCfg: CTestConfiguration, verifier: Verifier) =
        if (isValid) run(testCfg, verifier) else null

    private fun ExecutionScenario.run(testCfg: CTestConfiguration, verifier: Verifier): LincheckFailure? {
        val strategy = testCfg.createStrategy(testClass, this, testStructure.validationFunctions, verifier)
        return strategy.run()
    }

    private fun ExecutionScenario.copy() = ExecutionScenario(
        ArrayList(initExecution),
        parallelExecution.map { ArrayList(it) },
        ArrayList(postExecution)
    )

    private val ExecutionScenario.isValid: Boolean
        get() = !isParallelPartEmpty &&
                (!hasSuspendableActors() || (!hasSuspendableActorsInInitPart && !hasPostPartAndSuspendableActors))

    private fun ExecutionScenario.validate() {
        require(!isParallelPartEmpty) {
            "The generated scenario has empty parallel part"
        }
        if (hasSuspendableActors()) {
            require(!hasSuspendableActorsInInitPart) {
                "The generated scenario for the test class with suspendable methods contains suspendable actors in initial part"
            }
            require(!hasPostPartAndSuspendableActors) {
                "The generated scenario  for the test class with suspendable methods has non-empty post part"
            }
        }
    }

    private val ExecutionScenario.hasSuspendableActorsInInitPart get() =
        initExecution.stream().anyMatch(Actor::isSuspendable)
    private val ExecutionScenario.hasPostPartAndSuspendableActors get() =
        (parallelExecution.stream().anyMatch { actors -> actors.stream().anyMatch { it.isSuspendable } } && postExecution.size > 0)
    private val ExecutionScenario.isParallelPartEmpty get() =
        parallelExecution.map { it.size }.sum() == 0


    private fun CTestConfiguration.createVerifier() =
        verifierClass.getConstructor(Class::class.java).newInstance(sequentialSpecification).also {
            if (requireStateEquivalenceImplCheck) it.checkStateEquivalenceImplementation()
        }

    private fun CTestConfiguration.createExecutionGenerator() =
        generatorClass.getConstructor(
            CTestConfiguration::class.java,
            CTestStructure::class.java
        ).newInstance(this, testStructure)

    // This companion object is used for backwards compatibility.
    companion object {
        /**
         * Runs the specified concurrent tests. If [options] is null, the provided on
         * the testing class `@...CTest` annotations are used to specify the test parameters.
         *
         * @throws AssertionError if any of the tests fails.
         */
        @JvmOverloads
        @JvmStatic
        fun check(testClass: Class<*>, options: Options<*, *>? = null) {
            LinChecker(testClass, options).check()
        }
    }
}


/**
 * This is a short-cut for the following code:
 * ```
 *  val options = ...
 *  LinChecker.check(testClass, options)
 * ```
 */
fun <O : Options<O, *>> O.check(testClass: Class<*>) = LinChecker.check(testClass, this)

internal fun <O : Options<O, *>> O.checkImpl(testClass: Class<*>) = LinChecker(testClass, this).checkImpl()